go-controller/pkg/ovn/dns-name-resolver/egressfirewall_external_dns.go

```diff
@@ -0,0 +1,386 @@
+			addresses = append(addresses, resolvedAddress.IP)
+		}
+	}
+	addResp := extEgDNS.Add(AddRequest{DNSName: dnsName, Addresses: addresses})
+	return addResp.Err
+
+}
+
+// Add is called whenever a DNS name is needed to be added or updated.
+func (extEgDNS *ExternalEgressDNS) Add(request AddRequest) AddResponse {
+	extEgDNS.dnsLock.Lock()
+	defer extEgDNS.dnsLock.Unlock()
+
+	// Get the details of the resolved name corresponding to the DNS name.
+	resolvedName, exists := extEgDNS.dnsNames[request.DNSName]
+
+	// If the details of the resolved name is not available, then create
+	// the address set for the DNS name.
+	if !exists {
+		if extEgDNS.addressSetFactory == nil {
+			return AddResponse{
+				Err: fmt.Errorf("error adding DNS name %s: addressSetFactory is nil", request.DNSName),
+			}
+		}
+		asIndex := GetEgressFirewallDNSAddrSetDbIDs(request.DNSName, extEgDNS.controllerName)
+		dnsAddressSet, err := extEgDNS.addressSetFactory.NewAddressSet(asIndex, nil)
+		if err != nil {
+			return AddResponse{
+				Err: fmt.Errorf("cannot create AddressSet for DNS name %s: %v", request.DNSName, err),
+			}
+		}
+		resolvedName.dnsAddressSet = dnsAddressSet
+		resolvedName.namespaces = sets.New[string]()
+	}
+
+	// Update the address set of the DNS name with the IPs. The IPs which don't match the ip mode
+	// and those which match the cluster subnet are ignored.
+	var ips []net.IP
+	for _, addr := range request.Addresses {
+		ip := net.ParseIP(addr)
+		if ip != nil {
+			ips = append(ips, ip)
+		}
+	}
+
+	// Ignore ips from clusterSubnet, since this subnet shouldn't be affected by egress firewall.
+	filteredIPs := []net.IP{}
+	for _, ip := range ips {
+		ignoreIP := false
+
+		for _, clusterSubnet := range config.Default.ClusterSubnets {
+			if clusterSubnet.CIDR.Contains(ip) {
+				ignoreIP = true
+				break
+			}
+		}
+
+		if !ignoreIP {
+			filteredIPs = append(filteredIPs, ip)
+		}
+	}
+	if err := resolvedName.dnsAddressSet.SetIPs(filteredIPs); err != nil {
+		return AddResponse{
+			Err: fmt.Errorf("cannot add IPs to AddressSet for DNS name %s: %v", request.DNSName, err),
+		}
+	}
+
+	// Add the updated resolved name to the dnsNames map corresponding to the DNS name.
+	extEgDNS.dnsNames[request.DNSName] = resolvedName
+
+	return AddResponse{
+		Err: nil,
+	}
+}
+
+// Delete is called whenever a DNS name is needed to be deleted.
+func (extEgDNS *ExternalEgressDNS) Delete(request DeleteRequest) DeleteResponse {
+	extEgDNS.dnsLock.Lock()
+	defer extEgDNS.dnsLock.Unlock()
+
+	// Get the resolved name details corresponding to the DNS name.
+	resolvedName, exists := extEgDNS.dnsNames[request.DNSName]
+	// Return if the corresponding resolved name details is not found.
+	if !exists {
+		klog.Errorf("Details missing for DNS name %s. Skipping address set destroy operation.", request.DNSName)
+		return DeleteResponse{
+			Err: nil,
+		}
+	}
+
+	// Check if the DNS name is still in use in any namespace.
+	if resolvedName.namespaces.Len() != 0 {
```

Comment:
We may need to mark this addressset somehow to ensure we delete it when the last namespace dereferences it?

Reply:
This will be ensured by cluster-manager by deleting the DNSNameResolver object if the DNS name is not used in any namespace. When the object is deleted, then the corresponding address set will also be deleted.